### PR TITLE
feat(api): added caching system to views and viewsets

### PR DESCRIPTION
**Pull request by hamid-yg**
# Description

I'm adding a caching system to views and viewsets (Vulns, Missions...)

# Fixes

Specify which issue this PR targets, like so:

- feat #215 

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it.

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added @vrn-sh/members to the reviewers, or specific members of the team
- [x] I have added the needed labels
- [ ] I have tested this code
- [ ] I have added / updated tests (unit / functionals / end-to-end / ...)
- [ ] I have updated the README and other relevant documents (guides...)

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it.

### CODE REVIEW

**Comment by hamid-yg** [comment]
For which models should I add the caching system ?

File: api/api/tests/vuln_tests.py [hunk 1]
```diff
@@ -31,34 +32,41 @@ def tearDown(self) -> None:
         self.manager.delete()
         self.other_user.delete()
 
-    def test_create_valid_vulnerability(self):
```
Comment (djnnvx) [inline api/api/tests/vuln_tests.py:34]: is it not working ? why? :D

Comment (djnnvx) [reply api/api/tests/vuln_tests.py:34]: i can lookup a fix if you need help

File: api/api/views/viewsets/mission.py [hunk 2]
```diff
@@ -436,61 +439,70 @@ class MissionViewset(viewsets.ModelViewSet):  # pylint: disable=too-many-ancesto
...
+        if not mission:
+            mission = self.get_object()
+            serializer = self.get_serializer(mission)
+            mission = serializer.data
+            cache.set(cached_key, mission, self.CACHE_TIMEOUT)
+            return super().retrieve(request, *args, **kwargs)
+        return Response(mission)
+
+    def list(self, request, *args, **kwargs):
+        cached_key = f'{self.CACHE_KEY_PREFIX}list'
```
Comment (djnnvx) [inline api/api/views/viewsets/mission.py]: this code will leak missions to users that should not have access to them. 
You could fix that by:

1) filter mission by checking if user should have [access to them](https://github.com/vrn-sh/erp/blob/next/api/api/models/mission.py#L114) (you can add the `@property` decorator if that helps)

2) append the team's id to the cache key to make sure each cached version is unique for each teams

File: api/api/views/viewsets/vulns.py [hunk 3]
```diff
@@ -140,6 +137,31 @@ def list(self, request, *args, **kwargs):
...
+        if not vuln:
+            vuln = self.get_object()
+            serializer = self.get_serializer(vuln)
+            vuln = serializer.data
+            cache.set(cache_key, vuln, self.CACHE_TIMEOUT)
+            return super().retrieve(request, *args, **kwargs)
+        return Response(vuln)
+
+    def list(self, request, *args, **kwargs):
+        cache_key = f'{self.CACHE_KEY_PREFIX}list'
```
Comment (djnnvx) [inline api/api/views/viewsets/vulns.py]: my comment from before also applies here i think

**Review by djnnvx — COMMENTED** [review]
almost perfect, good job)

File: api/api/tests/vuln_tests.py (hunk 1 above)
Comment (hamid-yg) [reply api/api/tests/vuln_tests.py:34]: Yes, I'm getting "unknow vuln_type" and I don't know why. Check it if you can

**Comment by hamid-yg** [comment]
How can I get the `team id` from the `request.user` in the list view ? Please check it @bogdzn

**Review by djnnvx — APPROVED** [review]
perfect

**Comment by djnnvx** [comment]
> How can I get the `team id` from the `request.user` in the list view ? Please check it @bogdzn


To get a related `Team` object from the user's ID, you can first call [get_user_model()](https://github.com/vrn-sh/erp/blob/next/api/api/models/__init__.py#L214) and then filter all teams by [is_member](https://github.com/vrn-sh/erp/blob/next/api/api/models/__init__.py#L193)

**Pull request merged**